Update InstallVisualStudio.ps1 to support VS2022 preview

DamianEdwards opened the pull request:
Updates `InstalVisualStudio.ps1` to support VS2022 preview.

The NetCoreTools component and old Windows SDK versions have been removed from VS2022.

DamianEdwards committed:
Update InstallVisualStudio.ps1 to support VS2022 preview

dougbu commented on eng/scripts/InstallVisualStudio.ps1:
```diff
@@ -32,8 +36,10 @@
     [ValidateSet('BuildTools','Community', 'Professional', 'Enterprise')]
     [string]$Edition = 'Enterprise',
-    [ValidateSet('Release', 'Preview')]
+    [ValidateSet('Release', 'Preview', 'Preview')]
```
Needs less cowbell despite the beauty of https://www.youtube.com/watch?v=cVsQLlk-T0s
```suggestion
    [ValidateSet('Release', 'Preview')]
```

dougbu commented on eng/scripts/InstallVisualStudio.ps1:
```diff
@@ -59,29 +65,36 @@ mkdir $intermedateDir -ErrorAction Ignore | Out-Null
     if (Test-Path $vsWhere)
     {
-        $installations = & $vsWhere -version '[16,17)' -format json -prerelease -products $productId | ConvertFrom-Json |Sort-Object -Descending -Property installationVersion, installDate
+        $installations = & $vsWhere -version '[16,18)' -format json -prerelease -products $productId | ConvertFrom-Json |Sort-Object -Descending -Property installationVersion, installDate
```
This will mean 2022 will overwrite 2019. Suggest
```suggestion
        $installations = & $vsWhere -version "[$vsversion,$($vsversion+1))" -format json -prerelease -products $productId | ConvertFrom-Json |Sort-Object -Descending -Property installationVersion, installDate
```

DamianEdwards replied:
would need to be plus 2 yes, due to the `[ )` range limiter?

DamianEdwards replied:
nevermind, brain slow

dougbu replied:
No, current `[16,17)` works fine for 2019 and `[17,18}` will be fine for 2022

dougbu replied:
IOW I did the `pwsh` right 😀

dougbu commented on eng/scripts/InstallVisualStudio.ps1:
```diff
@@ -93,10 +106,16 @@ if (-not $InstallPath) {
 }
 
 if (-not $InstallPath) {
+    if ($vsversion -eq "16") {
+        $pathPrefix = "${env:ProgramFiles(x86)}";
+    }
```
🥇

dougbu reviewed (approved):
nitty nits

DamianEdwards committed:
Update eng/scripts/InstallVisualStudio.ps1

Co-authored-by: Doug Bunting <6431421+dougbu@users.noreply.github.com>

DamianEdwards committed:
Remove Windows 10 SDK component from global.json

Pilchie commented on global.json:
```diff
@@ -22,8 +22,7 @@
         "Microsoft.VisualStudio.Component.VC.ATL",
-        "Microsoft.VisualStudio.Component.VC.Tools.x86.x64",
-        "Microsoft.VisualStudio.Component.Windows10SDK.17134"
+        "Microsoft.VisualStudio.Component.VC.Tools.x86.x64"
```
I ran into this the other day - we are going to need *some* windows SDK I think, but that one no longer seems to be available in VS2022.

DamianEdwards replied:
Yeah, @dougbu said he'll follow up with the Arcade folks

dougbu replied:
Just confirmed `Microsoft.VisualStudio.Component.Windows10SDK.18362` (the new minimum in 2022) is available in 2019 too. I'm asking in the Arcade channel for that component to be added to the VS configuration on all agents in their next roll-out or perhaps the week after

dougbu replied:
/fyi https://teams.microsoft.com/l/message/19:1dad2081c8634f34915d88dce6220265@thread.skype/1626396237512?tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47&groupId=4d73664c-9f2f-450d-82a5-c2f02756606d&parentMessageId=1626396237512&teamName=.NET%20Core%20Eng%20Services%20Partners&channelName=Arcade&createdTime=1626396237512

Pilchie reviewed (approved):
Thanks for doing this!

DamianEdwards committed:
PR feedback

DamianEdwards commented:
Ran this locally on a new machine and can confirm `restore.cmd` completed successfully afterwards. However, build fails due to the Windows SDK issue:
```
PS C:\src\GitHub\dotnet\aspnetcore> .\eng\build.cmd -NoRestore -NoBuildJava
WARNING: No default group of projects was specified, so building the managed and native projects and their
dependencies. Run `build.cmd -help` for more details.
Building of C# project is enabled and has dependencies on NodeJS projects. Building of NodeJS projects is enabled since node is detected in C:\Program Files.

  RepoTasks -> C:\src\GitHub\dotnet\aspnetcore\artifacts\bin\RepoTasks\Release\net6.0\RepoTasks.dll
  RepoTasks -> C:\src\GitHub\dotnet\aspnetcore\artifacts\bin\RepoTasks\Release\net472\RepoTasks.dll

C:\Program Files\Microsoft Visual Studio\2022\Enterprise_Pre\MSBuild\Microsoft\VC\v170\Microsoft.Cpp.WindowsSDK.targets
(46,5): error MSB8036: The Windows SDK version 10.0.17134.0 was not found. Install the required version of Windows SDK
or change the SDK version in the project property pages or by right-clicking the solution and selecting "Retarget solut
ion". [C:\src\GitHub\dotnet\aspnetcore\src\Servers\IIS\AspNetCoreModuleV2\IISLib\IISLib.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise_Pre\MSBuild\Microsoft\VC\v170\Microsoft.Cpp.WindowsSDK.targets
(46,5): error MSB8036: The Windows SDK version 10.0.17134.0 was not found. Install the required version of Windows SDK
or change the SDK version in the project property pages or by right-clicking the solution and selecting "Retarget solut
ion". [C:\src\GitHub\dotnet\aspnetcore\src\Servers\IIS\AspNetCoreModuleV2\IISLib\IISLib.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise_Pre\MSBuild\Microsoft\VC\v170\Microsoft.Cpp.WindowsSDK.targets
(46,5): error MSB8036: The Windows SDK version 10.0.17134.0 was not found. Install the required version of Windows SDK
or change the SDK version in the project property pages or by right-clicking the solution and selecting "Retarget solut
ion". [C:\src\GitHub\dotnet\aspnetcore\src\Servers\IIS\AspNetCoreModuleV2\gtest\gtest.vcxproj]
C:\Program Files\Microsoft Visual Studio\2022\Enterprise_Pre\MSBuild\Microsoft\VC\v170\Microsoft.Cpp.WindowsSDK.targets
(46,5): error MSB8036: The Windows SDK version 10.0.17134.0 was not found. Install the required version of Windows SDK
or change the SDK version in the project property pages or by right-clicking the solution and selecting "Retarget solut
ion". [C:\src\GitHub\dotnet\aspnetcore\src\Servers\IIS\AspNetCoreModuleV2\gtest\gtest.vcxproj]

Build FAILED.
```

dougbu commented on eng/scripts/InstallVisualStudio.ps1:
```diff
@@ -59,29 +65,36 @@ mkdir $intermedateDir -ErrorAction Ignore | Out-Null
+    $Channel = "Preview";
+}
+$channelUri = "https://aka.ms/vs/$vsversion/release"
+$responseFileName = "vs.$vsversion"
```
vs.buildtools.json, vs.buildtools.preview.json, and vs.json files also need to be renamed

DamianEdwards replied:
Updated in a93a61c

dougbu reviewed (changes requested):
I missed something&hellip;

dougbu commented:
> Trying again with `-NoBuildNative`

Ick. Do we specify `17134` somewhere else in this repo❔

DamianEdwards commented:
@dougbu seems the native projects (e.g. IIS module) have to migrated to the new SDK?

DamianEdwards committed:
Update vs.buildtools.json for VS2022

dougbu reviewed (commented):
An off-by-`1` error: Still need to rename vs.buildtools.json. Then you're done

DamianEdwards commented:
Not being able to build ANCM is fairly crippling in my quick testing. I tried to build just the `src/ProjectTemplates` project with `build.cmd -pack -NoRestore -NoBuildNative` and it still fails 😢 . We could try multi-targeting the native projects to use the Windows SDK version that exists maybe?
```
Build FAILED.

C:\src\GitHub\dotnet\aspnetcore\src\Servers\IIS\IIS\src\Microsoft.AspNetCore.Server.IIS.csproj(32,5): error : Required dll from ANCM has not been built. To build ANCM, you must use MSBuild.exe.
    0 Warning(s)
    1 Error(s)
```

DamianEdwards committed:
Create vs.16.buildtools.json

DamianEdwards commented:
OK the various `vs.version.edition.channel.json` files should be there now but we still have the issue of the native projects not building.

dougbu commented:
> seems the native projects (e.g. IIS module) have to migrated to the new SDK?

Let's take this and the numerous `.vsconfig` files that list the specific Win10 SDK version separately and separate from each other. Up to you whether to include part 1 in this PR.

1. Make the .vcxproj logic conditional and use `10.0.17134.0` for 2019 and `10.0.18362.0` between now and when the CI agents can be updated.
2. Update the .vcconfig files after all of the 2019 agents have `Microsoft.VisualStudio.Component.Windows10SDK.18362`

dougbu committed:
Demo a .vcxproj change

- this should work 😺 
- suggest duplicating this in the other .vcxproj files if I'm right

dougbu commented:
Added a commit which should fix one .vcxproj&hellip;

DamianEdwards commented:
@dougbu I don't think I can test if that change works in isolation as that project depends on some of the other native projects. I'll try changing them in the same fashion to see if it works for the "only VS2022 installed" case at least.

dougbu commented:
> I'll try changing them in the same fashion to see if it works for the "only VS2022 installed" case at least.

Sorry, I changed the first .vcxproj that needed changing rather than the lowest-level one. src\Servers\IIS\AspNetCoreModuleV2\IISLib\IISLib.vcxproj may be at the bottom of the dependency graph. Maybe change that one and see if the errors move up to CommonLib.vcxproj❔

DamianEdwards committed:
Make native projects toolset version conditional for VS2022 support

DamianEdwards commented:
@dougbu I changed them all and then the next layer of the onion appear: PlatformToolset. So I made that conditional too. It now builds on my machine! We'd need to get some verification it still works properly on VS2019 dev boxes and the build agents of course too. In particular I'm not sure about the `Build.Settings` build in 'servers/IIS/build'.

DamianEdwards commented:
@dougbu seems the mac failure was transient? Can you rerun pls?

dougbu commented:
> Can you rerun pls?

You're most welcome
